http-client/src/main/java/io/airlift/http/client/JsonResponseHandler.java

```diff
@@ -91,7 +95,38 @@ public T handle(Request request, Response response)
         }
         catch (IllegalArgumentException e) {
             String json = new String(bytes, UTF_8);
-            throw new IllegalArgumentException(String.format("Unable to create %s from JSON response:\n[%s]", jsonCodec.getType(), json), e);
+            throw new IllegalArgumentException(format("Unable to create %s from JSON response:\n[%s]", jsonCodec.getType(), json), e);
         }
     }
+
+    private static String reportResponseContent(Response response)
+    {
+        byte[] bytes = new byte[1024];
+        int read;
+        try {
+            read = ByteStreams.read(response.getInputStream(), bytes, 0, bytes.length);
+        }
+        catch (IOException e) {
+            return "Error reading response from server";
+        }
+
+        if (read == 0) {
+            return "Response empty";
+        }
+
+        Charset assumedCharset = Optional.ofNullable(response.getHeader(CONTENT_TYPE))
+                .flatMap(input -> {
+                    try {
+                        return Optional.of(MediaType.parse(input));
+                    }
+                    catch (RuntimeException ignore) {
+                        return Optional.empty();
+                    }
+                })
+                .flatMap(mediaType -> mediaType.charset().toJavaUtil())
+                .orElse(UTF_8);
+
+        String interpretedResponse = new String(bytes, 0, read, assumedCharset);
+        return format("Response: [%s...] [%s...]", interpretedResponse, base16().encode(bytes, 0, read));
```

Comment:
why not base64?

Reply:
Always using `...` here is misleading, since it indicates truncation even when it is the full string. We could change this to do
```java
int MAX_CODE_POINTS = 1024;

// allow at least max Unicode code points in UTF-8
byte[] bytes = toByteArray(limit(response.getInputStream(), MAX_CODE_POINTS * 4));

// invalid input maps to charset's default replacement string
String interpretedResponse = new String(bytes, assumedCharset);

return format("Response: [%s] [%s]",
        Ascii.truncate(interpretedResponse, MAX_CODE_POINTS, "..."),
        Ascii.truncate(base16().encode(bytes, 0, read), MAX_CODE_POINTS * 2, "..."));